### DISCUSSION TITLE
How to run browser-side script after a route is rendered?

### DISCUSSION THREAD

**Discussion by humatic** [react-static/react-static#1588] [Q&A] [answered]
Generally speaking, how does one go about executing Javascript on the browser-side after particular routes are rendered? I have a feeling that there's a simple solution.

More specifically, my use-case is this... I'm attempting to use browser-side MathJax to render math symbols in individual blog posts in my app. This means that the first time a browser visits one of the blog posts that requires MathJax, the browser will request the relevant JS scripts from a CDN:

```javascript
<script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
<script id="MathJax-script" async src="https://cdn.jsdelivr.net/npm/mathjax@3.1.2/es5/tex-mml-chtml.js"></script>
```

From testing, I see that I can use `<Head>` to accomplish this. However, the problem I'm encountering is that once the MathJax scripts are loaded in the browser, subsequent navigation to different blog posts will cause errors. Why? Because the global `MathJax` object is already loaded in the browser (apparently those MathJax scripts can only be safely loaded once).

As far as I can tell, the solution is to load the MathJax scripts dynamically from the browser with some JS: If `MathJax` hasn't been loaded yet, request it; otherwise manually call `MathJax.typeset();` (causing MathJax to reevaluate the current document).

**At a bare minimum**, I need to be able to trigger `MathJax.typeset();` from the browser for particular routes.

Generally, I would hope that there's a non-hack way to execute JS on the browser-side for particular routes. But I'm not seeing it right now. Any pointers?

**Comment by humatic** [comment]
After more tinkering, I found that putting some raw JS inside `<Head>` inside my Post component kind-of works:
```javascript
function Post() {
  ...
  return (
    <div>
      <Content />
      <Head>
        <script id="UseMathJax">{`
          MathJax.typeset();
        `}</script>
      </Head>
    </div>
  )
};
```
But I don't understand why `<Head>` is required. If I just use the `<script>` tags without `<Head>`, then the JS won't execute. There is something I'm not understanding about how the static pages are interpreted by the browser.

**Reply by SleeplessByte** [reply]
In your `Post` route component you could imperatively call it:

```javascript
useLayoutEffect(() => {
  MathJax.typeset()
}, [])
```

**Reply by humatic** [reply]
I had to review the React Hooks docs, but I think I understand now, thanks! This works. But why use `useLayoutEffect` instead of `useEffect`?

**Reply by SleeplessByte** [reply]
So there is no flash of unstyled content :)

**Comment by humatic** [comment, marked as answer]
Thanks to @SleeplessByte I've got a solution. For posterity:

```js
import React, { useLayoutEffect } from 'react';

/*
  Include this component to dynamically load, configure, and apply MathJax to
  the current view. This component can be rendered from inside a Markdown
  document (mdx) or a normal React component.
*/
export default function UseMathJax() {

  useLayoutEffect(() => {
    if (typeof MathJax != 'undefined') {
      MathJax.typeset();
    } else if (!document.getElementById('MathJax-script')) {
      window.MathJax = {
        // http://docs.mathjax.org/en/latest/options/input/tex.html
        tex: {
          inlineMath: [
            ['\\(', '\\)'],
            // ['$', '$'],
          ],
          displayMath: [
            ['$$', '$$'],
            ['\\[', '\\]'],
          ],
        }
      };
      const script = document.createElement('script');
      script.src = 'https://cdn.jsdelivr.net/npm/mathjax@3.1.2/es5/tex-mml-chtml.js';
      script.async = true;
      script.id = 'MathJax-script';
      script.onload = () => { console.log('MathJax loaded') };
      document.head.appendChild(script);
    }
  }, []);

  return null;
};
```
